keps/sig-node/3545-improved-multi-numa-alignment/README.md

```diff
@@ -0,0 +1,444 @@
+
+## Design Details
+
+
+### Proposed Change
+ 
+We propose to
+- add a new flag in Kubelet called `TopologyManagerPolicyOptions` in the kubelet config or command line argument called `topology-manager-policy-options` which allows the user to specify the Topology Manager policy option.
+- add a new topology manager option called `prefer-closest-numa-nodes`; if present, this option will enable further refinements of the existing `restricted` and `best-effort` policies, this option has no effect for `none` and `single-numa-node` policies.
+
+Best-effort or restricted policy chooses resources that will fit into the smallest number of NUMA nodes.
+This enhancement wants to change how the best hint is chosen from choosing narrower `NUMANodeAffinity` bitmask to choosing the one that is the narrowest and has the shortest average distance between NUMA nodes.
+
+To summarize properties of `prefer-closest-numa-nodes` policy:
+
+* Preserve all properties of `best-effort` and `restricted` policies
+* Choose `NUMANodeAffinity` bitmask which is the narrowest and if the bitmask width is the same prefer the bitmask with smaller average distance.
+
+When `prefer-closest-numa-nodes` policy is enabled, we need to retrieve information regarding distances between NUMA nodes.
+Right now Topology manager discovers Node layout using [CAdvisor API](https://github.com/google/cadvisor/blob/master/info/v1/machine.go#L40).
+
+We would need to extend the `MachineInfo` struct with Distances field which would be an array of uint64 which will describe the distance between a given node and other nodes in the system.
+This information can be read from sysfs:
+
+```go 
+const     distanceFile = "distance"
+
+func (fs *realSysFs) GetDistance(nodePath string) (string, error) {
+    distancePath := fmt.Sprintf("%s/%s", nodePath, distanceFile)
+    distance, err := ioutil.ReadFile(distancePath)
+    if err != nil {
+        return "", err
+    }
+    return strings.TrimSpace(string(distance)), err
+}
+
+func getDistance(sysFs sysfs.SysFs, nodeDir string) ([]uint64, error) {
+    rawDistance, err := sysFs.GetDistance(nodeDir)
+    if err != nil {
+        //Ignore if per-node info is not available.
+        klog.Warningf("Found node without distance information, nodeDir: %s", nodeDir)
+        return nil, nil
+    }
+
+    distances := []uint64{}
+    for _, distance := range strings.Split(rawDistance, " ") {
+        distanceInt, err := strconv.ParseUint(distance, 10, 64)
+        if err != nil {
+            return nil, fmt.Errorf("cannot convert %s to int", distance)
+        }
+        distances = append(distances, distanceInt)
+    }
+
+    return distances, nil
+}
+```
+
+But before this lands into `cadvisor`, this information would have to be discovered by `kubelet`.
```

Comment:
how this will be handled? Will we have temporarily the same code above?

Reply:
Yes.